Make Terminal Logger respect verbosity

AR-May opened the pull request:
Fixes #9667, #9654

### Context
We would like to make the terminal logger partially respect verbosity.

### Changes Made
In case of Quiet verbosity, none of the static part of the Terminal Logger output, which is grouped by the project, is shown. Warnings and errors are shown as they come immediately.
In case of Minimal and Normal verbosity, the behavior stands intact.
In case of Detailed and Diagnostic verbosity, the terminal logger shows all the high priority messages in the static part of the output, under the corresponding project.

### Testing
Locally & unit tests

baronfel commented:
Is is possible to show renderings of some of the test 'verified' files/scenarios? It'd be good to be able to visually verify the changes here.

AR-May commented:
> Is is possible to show renderings of some of the test 'verified' files/scenarios? It'd be good to be able to visually verify the changes here.

Sure, thank you for the pointers how to do the rendering from the test verification files. 
Here are different verbosities levels for the same input (which consists of immediate message, messages with all verbosities, warning and error, all for one project):
Quiet:
![TerminalLogger_Tests PrintBuildSummaryQuietVerbosity_FailedWithErrors Windows verified](https://github.com/dotnet/msbuild/assets/67507805/97892e18-49f8-4767-b175-9323eb9f0bd4)

Minimal, Normal:
![TerminalLogger_Tests PrintBuildSummaryNormalVerbosity_FailedWithErrors Windows verified](https://github.com/dotnet/msbuild/assets/67507805/4518b07f-d4a6-42e8-8b39-3d99070b165d)

Detailed, Diagnostic:
![TerminalLogger_Tests PrintBuildSummaryDiagnosticVerbosity_FailedWithErrors Windows verified](https://github.com/dotnet/msbuild/assets/67507805/54df8cba-ecc1-4f3c-afc8-e46d0f2f7dfd)

JanKrivanek commented on src/MSBuild/TerminalLogger/TerminalLogger.cs:
```diff
@@ -212,8 +220,14 @@ internal TerminalLogger(ITerminal terminal)
+    {
+        get { return _verbosity; }
+        set { _verbosity = value; }
+    }
```
nit: do we need the backing field?

JanKrivanek commented on src/MSBuild/TerminalLogger/TerminalLogger.cs:
```diff
@@ -295,43 +309,46 @@ private void BuildFinished(object sender, BuildFinishedEventArgs e)
-            Terminal.WriteLine("");
-            if (_restoreFailed)
+        { 
+            if (Verbosity > LoggerVerbosity.Quiet)
```
Do we need the `BeginUpdate` and the finally block for the quiet mode? Or can we extract clearing of the stats into a  method - and for quiet just call that and return? (less nested method would get bit more readable)

AR-May replied:
We need the finally block to execute there, and it needs calling of `BeginUpdate`. We might do as you suggest: put the finally block and clearing in a separate method, but I thought it would be easier to add the extra if condition to the messages I want to skip.

JanKrivanek commented on src/MSBuild/TerminalLogger/TerminalLogger.cs:
```diff
@@ -400,166 +417,170 @@ private void ProjectFinished(object sender, ProjectFinishedEventArgs e)
-
-        if (_projects.TryGetValue(c, out Project? project))
+        // Add project summary to the static part of the Console only if verbosity is higher than Quiet.
+        if (Verbosity > LoggerVerbosity.Quiet)
```
Can we actually just shortcut quit the mothod for quiet?
It would get the diff more readable

JanKrivanek replied:
Or group the condition with the existing one

AR-May replied:
I agree, shortcut quit is better here.

JanKrivanek reviewed (commented):
It'd be nice to short-circuit return from the methods on quiet mode - so that diff would get less cluttered

JanKrivanek reviewed (approved):
Looks good!

Do you want to flip to "ready for review"?

AR-May commented:
I wonder should we put this change under a change wave?

baronfel commented:
IMO no - terminal logger is a live, always updating feature

AR-May commented:
I ran the TL with detailed verbosity on multiple repositories to check the behavior. All seems to be fine, except for one very annoying line that shows on detailed and diagnostics verbosity. It is a line produced by "exec" task and it shows the exec command line for the compiler call as [High priority message](https://github.com/dotnet/msbuild/blob/de776177f6d540e656e6b0c6d5bb07f2ff518c19/src/Utilities/ToolTask.cs#L477). The message is usually huge. The console logger has a work-around that allows not to show this line when minimal verbosity is used. Maybe not for this PR, but should we consider implementing similar workaround in TL or fixing the priority of this message?

baronfel commented:
>Maybe not for this PR, but should we consider implementing similar workaround in TL or fixing the priority of this message?

Yes, we should detect + workaround this. I'm not sure we can change the priority, maybe there is some other flag/signal we can set or detect?

Now that we have forward-compatibility this could almost be a specific message type :D

AR-May commented:
> > Maybe not for this PR, but should we consider implementing similar workaround in TL or fixing the priority of this message?
> 
> Yes, we should detect + workaround this. I'm not sure we can change the priority, maybe there is some other flag/signal we can set or detect?
> 
> Now that we have forward-compatibility this could almost be a specific message type :D


Well, we have a specific message type `TaskCommandLineEventArgs` and that is how it is [detected in the console logger](https://github.com/dotnet/msbuild/blob/de776177f6d540e656e6b0c6d5bb07f2ff518c19/src/Build/Logging/ParallelLogger/ParallelConsoleLogger.cs#L1136) and skipped. Although the console logger has also a specific parameter that allows for it to be shown, and I suggest having it too. In this case it would be easier for me to implement it in the second PR concerning verbosity where I add parsing of /tlp. @baronfel do we mind if some of MSBuild versions (after merging this PR and before merging my next PR concerning the verbosity) will write this annoying line for a detailed verbosity?

AR-May commented:
Adding the parameter to terminal logger parameters appeared to be easier (apparently, we already have most of the code in place), so I added it in this PR. 
So, the changes:
- parameter `/tlp:v=...` will overwrite the verbosity of the TL that is set by `/v` parameter. In case of bad parameter value, it throws the error, like `/clp:v=...` does.
- I added a parameter `/tlp:SHOWCOMMANDLINE` for opting-in the `TaskCommandLineEventArgs` messages, similar to the `/clp:SHOWCOMMANDLINE` parameter. By default, we will skip `TaskCommandLineEventArgs` messages on detailed and diagnostic verbosity of TL, as the console logger does so on minimal verbosity. In case of bad parameter value does not throw an error, just ignores it, like `/clp` does.

@JanKrivanek, @rokonec could you please review couple of latest commits? 
@baronfel is the change and workaround for opting the line back in fine for you?

baronfel commented:
@AR-May that seems very acceptable to me, thank you!

JanKrivanek commented on src/Build/Logging/BaseConsoleLogger.cs:
```diff
@@ -1038,33 +1030,17 @@ internal virtual bool ApplyParameter(string parameterName, string parameterValue
+                Verbosity = (LoggerVerbosity)verbosity;
+                return true;
+            }
+            else
+            {
+                string errorCode;
+                string helpKeyword;
+                string message = ResourceUtilities.FormatResourceStringStripCodeAndKeyword(out errorCode, out helpKeyword, "InvalidVerbosity", parameterValue);
+                throw new LoggerException(message, null, errorCode, helpKeyword);
             }
```
This code seems repeated. Should we rename `TryParseVerbosityParameter` to `ParseVerbosityParameter` and have it throw the exception if it is not able to parse?

AR-May replied:
Eventually I hope we will remove all the shared files, but I feel this work is a bit out of scope for this PR.

JanKrivanek commented on src/MSBuild/TerminalLogger/TerminalLogger.cs:
```diff
@@ -255,6 +268,83 @@ public void Initialize(IEventSource eventSource)
+        }
+        else
+        {
+            try
+            {
+                _showCommandLine = ConversionUtilities.ConvertStringToBool(parameterValue);
+            }
+            catch (ArgumentException)
+            {
+                return false;
+            }
```
```suggestion
            return ConversionUtilities.TryConvertStringToBool(parameterValue, out _showCommandLine );
```

JanKrivanek reviewed (approved):
Looks good.
Added some suggestions for improvements

AR-May commented on src/MSBuild/TerminalLogger/TerminalLogger.cs:
```diff
@@ -255,6 +268,76 @@ public void Initialize(IEventSource eventSource)
+                return ApplyVerbosityParameter(parameterValue);
+            case "SHOWCOMMANDLINE":
+                return ApplyShowCommandLineParameter(parameterValue);
+        }
```
That's question for @baronfel. Should we? We do not aggressively fail for Console Logger, so I continued as such for Terminal Logger. But if we want to change it, sure, I can change it.

AR-May replied:
Note: we do not aggressively fail in most of the cases, but when verbosity is set wrongly, we do.

baronfel replied:
Failing on unknown seems like a fragile approach to me - imagine someone flipping between different SDKs where we add new functionality in version N+1 - the user should be able to use a response file or equivalent mechanism to hardcode that configuration without having to keep flipping that configuration on/off.

So - I think silently not applying is ok here.

AR-May commented on src/Framework/Logging/LoggerParametersHelper.cs:
```diff
@@ -0,0 +1,73 @@
+
+        public static IEnumerable<KeyValuePair<string, string?>> ParseParameters(string? parametersString)
+        {
+            List<KeyValuePair<string, string?>> parameters = new();
```
Well, why indeed! My bad, going to fix it. Also, I was not aware that there was a difference in perf related to my pick of KVP vs tuple. Intent-wise KVP fits better than tuple, but that's it. I agree, tuple is better since we are going to pass it as an argument.

AR-May commented on src/MSBuild/TerminalLogger/TerminalLogger.cs:
```diff
@@ -255,6 +268,76 @@ public void Initialize(IEventSource eventSource)
+    /// Apply a terminal logger parameter.
+    /// parameterValue may be null, if there is no parameter value.
+    /// </summary>
+    private bool ApplyParameter(string parameterName, string? parameterValue)
```
The value meant whether the function was successful or not. Should have called it `TryApplyParameter`.
The idea was not to throw on parse error and communicate the success this way. But since we anyway ignore this value, the class is sealed and the function is not public, we can change it anytime if we need such info for our logic, so I will just remove this bool.

AR-May commented on src/MSBuild/TerminalLogger/TerminalLogger.cs:
```diff
@@ -744,7 +869,9 @@ private void WarningRaised(object sender, BuildWarningEventArgs e)
 
-        if (buildEventContext is not null && _projects.TryGetValue(new ProjectContext(buildEventContext), out Project? project))
+        if (buildEventContext is not null
+            && _projects.TryGetValue(new ProjectContext(buildEventContext), out Project? project)
```
Good idea!

AR-May replied:
wrong chat :)

rainersigwald commented on src/Framework/Logging/LoggerParametersHelper.cs:
```diff
@@ -0,0 +1,69 @@
+            }
+        }
+
+        public static IEnumerable<Tuple<string, string?>> ParseParameters(string? parametersString)
```
Ah, I'd actually think about using a language tuple with named fields here but this is WAY into nit territory:

```suggestion
        public static IEnumerable<(string Name, string? Value)> ParseParameters(string? parametersString)
```

AR-May replied:
I will take this syntax as possibility next time, thanks!
